### PR TITLE
Fix : Get dimension length kind and Array size kind compatible

### LINKED ISSUE #5646: Bug : calculating the size for array based on kind=8 variable fails
MRE 
```fortran
PROGRAM subroutines_12

    IMPLICIT NONE
    
    INTEGER(8) :: nang = 2
    
    CALL my_subroutine
    
    CONTAINS
    
        SUBROUTINE my_subroutine
    
            INTEGER, DIMENSION(nang) :: fxhv
    
        END SUBROUTINE my_subroutine
    
    END PROGRAM subroutines_12
    
```

```console
; ModuleID = 'LFortran'
source_filename = "LFortran"

@subroutines_12.nang = internal global i64 2

define i32 @main(i32 %0, i8** %1) {
.entry:
  call void @_lpython_call_initial_functions(i32 %0, i8** %1)
  call void @my_subroutine()
  call void @_lpython_free_argv()
  br label %return

return:                                           ; preds = %.entry
  ret i32 0
}

define void @my_subroutine() {
.entry:
  %0 = load i64, i64* @subroutines_12.nang, align 4
  %1 = mul i32 1, i64 %0
  %2 = mul i32 %1, 4
  %3 = call i8* @_lfortran_malloc(i32 %2)
  %4 = bitcast i8* %3 to i32*
  br label %return

return:                                           ; preds = %.entry
  call void @_lfortran_free(i8* %3)
  ret void
}

declare i8* @_lfortran_malloc(i32)

declare void @_lfortran_free(i8*)

declare void @_lpython_call_initial_functions(i32, i8**)

declare void @_lpython_free_argv()
asr_to_llvm: module failed verification. Error:
Both operands to a binary operator are not of the same type!
  %1 = mul i32 1, i64 %0

code generation error: asr_to_llvm: module failed verification. Error:
Both operands to a binary operator are not of the same type!
  %1 = mul i32 1, i64 %0



Note: Please report unclear, confusing or incorrect messages as bugs at
https://github.com/lfortran/lfortran/issues.

```

### PR DESCRIPTION
**Pull request by assem2002**
fixes https://github.com/lfortran/lfortran/issues/5647
fixes  #5646

### CODE REVIEW

**Review by certik — APPROVED** [review]
As a temporary fix this is fine, however can you open up an issue and document there how to fix this properly and link it from the source code?

We need to start reducing this technical debt, so let's at least have it documented.

I think the ASR itself should have this figured out (either all 32 bit, or all 64 bit, and explicit casts in ASR itself).

**Comment by assem2002** [comment]
I don't get what actually to get documented, is that we don't support array size of int64 in llvm `array_descriptor`?
If it's about ASR node `ArraySize` , I think we can't cast this explicitly in the ASR the node looks like this :
```clj
(ArraySize
  (Var 3 fxhv)
  ()
  (Integer 2)
  ()
)
```

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 1]
```diff
@@ -3724,6 +3724,16 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
                 ptr_loads = 2;
                 for( size_t i = 0; i < n_dims; i++ ) {
                     this->visit_expr_wrapper(m_dims[i].m_length, true);
+
+                    // Make dimension length and return size compatible.(TODO : array_size should be of type int64)
+                    if(ASRUtils::extract_kind_from_ttype_t(
+                        ASRUtils::expr_type(m_dims[i].m_length)) > 32){
```
Comment (certik) [inline src/libasr/codegen/asr_to_llvm.cpp:3730]:
```suggestion
                        ASRUtils::expr_type(m_dims[i].m_length)) > 4){
```

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 2]
```diff
@@ -3724,6 +3724,16 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
...
+                        tmp = builder->CreateTrunc(tmp, llvm::IntegerType::get(context, 32));
+                    } else if (ASRUtils::extract_kind_from_ttype_t(
+                        ASRUtils::expr_type(m_dims[i].m_length)) < 32){
```
Comment (certik) [inline src/libasr/codegen/asr_to_llvm.cpp:3733]:
```suggestion
                        ASRUtils::expr_type(m_dims[i].m_length)) < 4){
```

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 3]
```diff
@@ -3724,6 +3724,16 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
                 ptr_loads = 2;
                 for( size_t i = 0; i < n_dims; i++ ) {
                     this->visit_expr_wrapper(m_dims[i].m_length, true);
+
+                    // Make dimension length and return size compatible.(TODO : array_size should be of type int64)
```
Comment (certik) [inline src/libasr/codegen/asr_to_llvm.cpp:3728]: Create an issue an link it here

Comment (assem2002) [reply src/libasr/codegen/asr_to_llvm.cpp:3728]: opened one https://github.com/lfortran/lfortran/issues/5679

File: src/libasr/codegen/asr_to_llvm.cpp (hunk 2 above)
Comment (assem2002) [reply src/libasr/codegen/asr_to_llvm.cpp]: done here https://github.com/lfortran/lfortran/pull/5648/commits/d19a917f94b9aa4ea76638914ab51aad2181cbaa

File: src/libasr/codegen/asr_to_llvm.cpp (hunk 1 above)
Comment (assem2002) [reply src/libasr/codegen/asr_to_llvm.cpp]: done here https://github.com/lfortran/lfortran/pull/5648/commits/d19a917f94b9aa4ea76638914ab51aad2181cbaa

File: src/libasr/codegen/asr_to_llvm.cpp (hunk 3 above)
Comment (certik) [reply src/libasr/codegen/asr_to_llvm.cpp:3728]: You should link it from the source code too.

**Review by certik — APPROVED** [review]

**Pull request merged**